### PR TITLE
Add support for the new download mechanism used in Unifi OS 4

### PR DESCRIPTION
**Pull request by cyberpower678**
### Description of change

- Adds export_camera_video() which is the exact implementation of what get_camera_video() used to be.
- Adds prepare_camera_video() which is the new implementation used by Unifi Protect 4, and the recommended way to get videos.
- Adds download_camera_video() which is the follow call that needs to be made after prepare_camera_video()
- Rewrites get_camera_video() to automatically choose between export or prepare/download to maintain backwards compatability.

### Pull-Request Checklist

- [ x ] Code is up-to-date with the `main` branch
- [ x ] This pull request follows the [contributing guidelines](https://github.com/uilibs/uiprotect/blob/main/CONTRIBUTING.md).
- [ ] This pull request links relevant issues as `Fixes #0000` (N/A) (export fails frequently when used on new versions of Protect)
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ x ] The new commits follow conventions outlined in the [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/), such as "fix(api): prevent racing of requests".

> - If pre-commit.ci is failing, try `pre-commit run -a` for further information.
> - If CI / test is failing, try `poetry run pytest` for further information.

### CODE REVIEW

**Comment by bdraco** [comment]
Thanks for the PR, I'll add some comments to get the review started shortly

File: src/uiprotect/api.py [hunk 1]
```diff
@@ -1451,25 +1451,152 @@ async def _stream_response(
             if progress_callback is not None:
                 await progress_callback(step, current, total)
 
-    async def get_camera_video(
+    async def download_camera_video(
+        self,
+        camera_id: str,
+        filename: str,
+        output_file: Optional[Path] = None,
```
Comment (bdraco) [inline src/uiprotect/api.py:1458]: Please use pep 0563 style

pyupgrade can likely convert these for you

https://peps.python.org/pep-0563/

File: src/uiprotect/api.py [hunk 2]
```diff
@@ -1538,6 +1666,64 @@ async def callback(total: int, chunk: bytes | None) -> None:
...
+                channel_index=channel_index,
+                validate_channel_id=validate_channel_id,
+                fps=fps,
+                filename=filename,
+            )
+
+            if isinstance(prepare_response, dict):
+                download_filename = prepare_response["fileName"]
+            else:
+                raise Exception
```
Comment (bdraco) [inline src/uiprotect/api.py:1703]: Please avoid using a broad exception raise in a try block to control flow

File: src/uiprotect/api.py [hunk 3]
```diff
@@ -1538,6 +1666,64 @@ async def callback(total: int, chunk: bytes | None) -> None:
...
+
+            return await self.download_camera_video(
+                camera_id=camera_id,
+                filename=download_filename,
+                output_file=output_file,
+                iterator_callback=iterator_callback,
+                progress_callback=progress_callback,
+                chunk_size=chunk_size,
+            )
+        except Exception:
```
Comment (bdraco) [inline src/uiprotect/api.py:1713]: Please avoid a broad exception catch and only catch exceptions that would confirm the need to fallback to the other method as it can unexpectedly hide bugs

**Comment by cyberpower678** [comment]
Let me conform it first to the way you have the typing handled.  :-)

File: src/uiprotect/api.py [hunk 4]
```diff
@@ -1451,25 +1451,152 @@ async def _stream_response(
             if progress_callback is not None:
                 await progress_callback(step, current, total)
 
-    async def get_camera_video(
+    async def download_camera_video(
```
Comment (bdraco) [inline src/uiprotect/api.py:1454]: Please add some tests for the new methods

**Review by bdraco — COMMENTED** [review]
Thanks for the PR. Please see feedback above.

File: src/uiprotect/api.py (hunk 1 above)
Comment (bdraco) [reply src/uiprotect/api.py:1458]: https://github.com/asottile/pyupgrade

File: src/uiprotect/api.py (hunk 3 above)
Comment (cyberpower678) [reply src/uiprotect/api.py:1713]: This was quite intentional, as it's supposed to fall back on the previous video/export on any kind of error, including using it on a pre-4 Protect installation whose responses I can't predict since I have no platform to test it on.  The get_camera_video() needs to be deprecated in favor of either using the export_camera_video method, which is functionally identical, or the new prepare/download_camera_video methods used in Protect 4.

**Comment by bdraco** [comment]
When I get back home I'll adjust the settings so you don't have to wait for the ci to rerun

File: src/uiprotect/api.py (hunk 3 above)
Comment (bdraco) [reply src/uiprotect/api.py:1713]: It's better for it to fail than to have undefined behavior since if it falls back without warning the problem will never get fixed, and when the deprecated function is removed we will find out the hard way when it breaks which will be harder to fix.

Please narrow the exception catch to only exceptions that are expected. 

https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/broad-exception-caught.html

Comment (bdraco) [reply src/uiprotect/api.py:1713]: This also needs a test to simulate a failure so we know under which conditions it should fallback to the old api

Comment (bdraco) [reply src/uiprotect/api.py:1713]: It should also log a message when it falls back so we know it's happening so we have a chance to fix whatever is causing the failure

**Comment by bdraco** [comment]
> When I get back home I'll adjust the settings so you don't have to wait for the ci to rerun 

I was able to adjust the settings on mobile so you shouldn't have to wait for manual ci runs 👍

**Comment by cyberpower678** [comment]
Sorry, but it's hard to follow our conversation with code rabbit producing a lot verbiage.  I'm not quite understanding the Lint Commit Messages thing.

**Comment by bdraco** [comment]
> Sorry, but it's hard to follow our conversation with code rabbit producing a lot verbiage.  I'm not quite understanding the Lint Commit Messages thing.

Don't worry about the commits as only the title matters since it will get squashed at the end anyways

**Comment by cyberpower678** [comment]
Alright, the pyunifiprotect library went completely poof, so I'll be focusing 100% of my attention here now. :p  Not sure what prompted the spontaneous deletion of the Christopher's fork of the library, but I see it was forked from you, which was forked from briis.  Just wondering what the story is if you know/willing to share.  I think it's safe to say that this is the official (unofficial) library for integrating with Unifi Protect now.

**Comment by bdraco** [comment]
> Alright, the pyunifiprotect library went completely poof, so I'll be focusing 100% of my attention here now. :p Not sure what prompted the spontaneous deletion of the Christopher's fork of the library, but I see it was forked from you, which was forked from briis. Just wondering what the story is if you know/willing to share. I think it's safe to say that this is the official (unofficial) library for integrating with Unifi Protect now.

This [comment](https://github.com/home-assistant/core/issues/117100#issuecomment-2121186460) best summarizes the situation

File: .gitignore [hunk 5]
```diff
@@ -184,3 +184,7 @@ event.json
 test_websocket.py
 test_raw.py
 rawdata.json
+
```
Comment (bdraco) [inline .gitignore:187]: Please do this in a separate PR as its not related to the new feature being added

File: .idea/uiprotect.iml [hunk 6]
```diff
@@ -1,9 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<module type="PYTHON_MODULE" version="4">
```
Comment (bdraco) [inline .idea/uiprotect.iml:2]: Please do the removal in another separate PR as its not related to the new feature being added

File: .gitignore (hunk 5 above)
Comment (cyberpower678) [reply .gitignore:187]: Sorry about that.  Still getting my environment organized, and meant to push this to a different branch.  I will fix that.

Comment (bdraco) [reply .gitignore:187]: Its no problem to add these if you do want to do a separate PR for them.

**Comment by bdraco** [comment]
I've been keeping this up to date with main.

No rush, but if you want to get this change in before HA beta on Wednesday, it will need another turn before Monday.

**Comment by cyberpower678** [comment]
> I've been keeping this up to date with main.
> 
> No rush, but if you want to get this change in before HA beta on Wednesday, it will need another turn before Monday.

I'll see what I can do.  My time has been very limited these past few weeks and will continue to remain limited until late July.  What's throwing me are the tests, since I can't quite seem to figure out the flow of logic that is letting make the proper tests for it.  I've been predominantly developing in PHP and have never dabbled with pytest before.

**Comment by bdraco** [comment]
> > I've been keeping this up to date with main.
> > No rush, but if you want to get this change in before HA beta on Wednesday, it will need another turn before Monday.
> 
> I'll see what I can do. My time has been very limited these past few weeks and will continue to remain limited until late July. What's throwing me are the tests, since I can't quite seem to figure out the flow of logic that is letting make the proper tests for it. I've been predominantly developing in PHP and have never dabbled with pytest before.

No worries. Take your time.

**Comment by ep1cman** [comment]
Has there been any progress on this? Its blocking unifi-protect-backup from being swapped to this library. If there is no plan to progress this PR, I might have to drop support for this new download method in order to make the switch

**Comment by bdraco** [comment]
@ep1cman If you want, I can give you access to this repo so you can finish this PR

**Comment by cyberpower678** [comment]
> Has there been any progress on this? Its blocking unifi-protect-backup from being swapped to this library. If there is no plan to progress this PR, I might have to drop support for this new download method in order to make the switch

So sorry, I just have not had any time.  I anticipate my time freeing up towards the end of July where I can then focus on this once again.  The only thing actually holding up this PR, is implementing some test cases, which I'm not too familiar with in Python.  If you know how to set those up, that would get this PR over the finish line.

**Comment by ep1cman** [comment]
Hi, I have some time this weekend and would like to get this PR over the line and merged as it is preventing unifi-protect-backup from fully making the switch to this library without losing functionality. Could you provide details of what is needed to get this merged?

**Comment by bdraco** [comment]
> Hi, I have some time this weekend and would like to get this PR over the line and merged as it is preventing unifi-protect-backup from fully making the switch to this library without losing functionality. Could you provide details of what is needed to get this merged?

Hi @ep1cman 

My outstanding review comments and the missing tests are the key blockers.

The bot review comments are not generally not blocking and sometimes have some good insights

**Comment by ep1cman** [comment]
Can you give me write permissions to this PR, or should I make a new one?

I believe I have addressed all the comments from yourself and the bot, except one. I disagree with the bot on downloading to a temp file if one isn't provided. The code already has logic to  write both to a file, or to memory. If you wish to use a temp file you can give it one. I would quite like to maintain the ability to not write files to disk as something like unifi-protect-backup can generate excessive (and unnecessary) disk IO otherwise.

I just need to write tests for this. Any advice on this would be welcome. I'm unsure of how to test this functionality without access to a unifi protect instance.

**Comment by bdraco** [comment]
You should have write access now

**Comment by ep1cman** [comment]
> You should have write access now

```
uiprotect on  heads/cyberpower678/main [⇕!] is 🐳 v24.0.9 via ·uiprotect-461Fq3yj-py3.12 
➜ git push
Enumerating objects: 58, done.
Counting objects: 100% (58/58), done.
Delta compression using up to 8 threads
Compressing objects: 100% (42/42), done.
Writing objects: 100% (52/52), 5.90 KiB | 2.95 MiB/s, done.
Total 52 (delta 33), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (33/33), completed with 5 local objects.
To github.com:cyberpower678/uiprotect.git
 ! [remote rejected] cyberpower678/main -> cyberpower678/main (permission denied)
error: failed to push some refs to 'github.com:cyberpower678/uiprotect.git'
(uiprotect-py3.12) 
uiprotect on  heads/cyberpower678/main [⇕!] is 🐳 v24.0.9 via ·uiprotect-461Fq3yj-py3.12 
➜ git push --force
Enumerating objects: 58, done.
Counting objects: 100% (58/58), done.
Delta compression using up to 8 threads
Compressing objects: 100% (42/42), done.
Writing objects: 100% (52/52), 5.90 KiB | 1.97 MiB/s, done.
Total 52 (delta 33), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (33/33), completed with 5 local objects.
To github.com:cyberpower678/uiprotect.git
 ! [remote rejected] cyberpower678/main -> cyberpower678/main (permission denied)
error: failed to push some refs to 'github.com:cyberpower678/uiprotect.git'
(uiprotect-py3.12) 
```

**Comment by bdraco** [comment]
Adjusted a few more settings, please try again

**Comment by ep1cman** [comment]
> Adjusted a few more settings, please try again

Sadly still no joy. I've had this work as a repo owner before but I dont think ive ever tried editing someone elses PR as a non-owner. Will it just be easier to open a new PR?

**Comment by bdraco** [comment]
> > Adjusted a few more settings, please try again
> 
> Sadly still no joy. I've had this work as a repo owner before but I dont think ive ever tried editing someone elses PR as a non-owner. Will it just be easier to open a new PR?

Yeah I think thats the way to go
